src/Microsoft.Health.Fhir.SqlServer/Features/Schema/SchemaInitializer.cs

```diff
@@ -0,0 +1,143 @@
+
+        public SchemaInitializer(SqlServerDataStoreConfiguration sqlServerDataStoreConfiguration, SchemaUpgradeRunner schemaUpgradeRunner, SchemaInformation schemaInformation, ILogger<SchemaInitializer> logger)
+        {
+            EnsureArg.IsNotNull(sqlServerDataStoreConfiguration, nameof(sqlServerDataStoreConfiguration));
+            EnsureArg.IsNotNull(schemaUpgradeRunner, nameof(schemaUpgradeRunner));
+            EnsureArg.IsNotNull(schemaInformation, nameof(schemaInformation));
+            EnsureArg.IsNotNull(logger, nameof(logger));
+
+            _sqlServerDataStoreConfiguration = sqlServerDataStoreConfiguration;
+            _schemaUpgradeRunner = schemaUpgradeRunner;
+            _schemaInformation = schemaInformation;
+            _logger = logger;
+        }
+
+        private void Initialize()
+        {
+            if (!CanInitialize())
+            {
+                return;
+            }
+
+            if (!IsSchemaTableInitialized())
+            {
+                CreateSchemaTable();
+            }
+
+            SetCurrentSchemaVersion();
+
+            if (_schemaInformation.Current == null || _schemaInformation.Current < _schemaInformation.Min)
+            {
+                int current = (int?)_schemaInformation.Current ?? 0;
+
+                for (int i = current + 1; i <= (int)_schemaInformation.Min; i++)
+                {
+                    _schemaUpgradeRunner.ApplySchema(i);
+                }
+            }
+        }
+
+        private void CreateSchemaTable()
+        {
+            using (var connection = new SqlConnection(_sqlServerDataStoreConfiguration.ConnectionString))
+            {
+                var command = new SqlCommand("CREATE TABLE SchemaTable ( Version int PRIMARY KEY, Status varchar(10) )", connection);
+
+                connection.Open();
+                command.ExecuteNonQuery();
+            }
+        }
+
+        private bool IsSchemaTableInitialized()
+        {
+            using (var connection = new SqlConnection(_sqlServerDataStoreConfiguration.ConnectionString))
+            {
+                var command = new SqlCommand("SELECT count(*) FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_NAME = 'SchemaTable'", connection);
+
+                connection.Open();
+                var tableCount = command.ExecuteScalar() as int?;
+
+                return tableCount == 1;
+            }
+        }
+
+        private void SetCurrentSchemaVersion()
+        {
+            using (var connection = new SqlConnection(_sqlServerDataStoreConfiguration.ConnectionString))
+            {
+                var command = new SqlCommand("SELECT MAX(Version) FROM SchemaTable WHERE Status = 'complete'", connection);
+
+                connection.Open();
+                int? version = null;
+
+                try
+                {
+                    version = command.ExecuteScalar() as int?;
+                }
+                catch (SqlException ex)
```

Comment:
Not a big fan of catching and moving on, since then we assume that the version must be 0 and we re-create the database from scratch.